### PR TITLE
Datanode/opensearch configuration beans

### PR DESCRIPTION
**Pull request by todvora**
This opensearch configuration handling continues in the direction https://github.com/Graylog2/graylog2-server/pull/21031 started. 

The very same configuration handling pattern will be used for Trino. 

image

Benefits of this handling:

- Repeatable configuration builds, groundwork for immutable and read-only configuration
- All configuration options are going through the same pipeline and data types => predictable code and locations instead of chaotic configuration through the whole datanode codebase.
- Configuration is written to temp directory, for each process run separately and independently (removes existing concurrency issues when restarting opensearch process)
- Unified opensearch configuration files handling 
- OpensearchConfigSync preflight check removed and replaced by config bean and config files handling
- Massively simplified certificates and security handling
- Unified approach for JVM options, opensearch.yml properties and opensearch.keystore entries
- Password secret completely removed from opensearch configuration, replaced by OTP
- Removes configuration side-effects from opensearch process impl

/nocl

## How Has This Been Tested?
Existing set of unit and integration tests

## Screenshots (if appropriate):

image

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: data-node/src/main/java/org/graylog/datanode/opensearch/OpensearchProcessImpl.java [hunk 1]
```diff
@@ -225,23 +207,11 @@ private void configure() {
...
-                    writeSeedHostsList();
                 }),
                 () -> {throw new IllegalArgumentException("Opensearch configuration required but not supplied!");}
         );
     }
 
-    private void writeSeedHostsList() {
-        try {
-            final Path hostsfile = datanodeConfiguration.datanodeDirectories().createOpensearchProcessConfigurationFile(UNICAST_HOSTS_FILE);
-            final Set<String> current = nodeService.allActive().values().stream().map(DataNodeDto::getClusterAddress).filter(Objects::nonNull).collect(Collectors.toSet());
```
Comment (moesterheld) [inline data-node/src/main/java/org/graylog/datanode/opensearch/OpensearchProcessImpl.java:238]: nodeService can be removed from class

**Review by moesterheld — CHANGES_REQUESTED** [review]
Awesome work. I haven't finished thorough testing of the PR yet, but here are already some code comments.

File: data-node/src/main/java/org/graylog/datanode/bindings/OpensearchProcessBindings.java [hunk 2]
```diff
@@ -49,9 +57,21 @@ protected void configure() {
...
+        opensearchCertificatesProviders.addBinding().to(LocalConfigurationCertificatesProvider.class).asEagerSingleton();
+        opensearchCertificatesProviders.addBinding().to(DatanodeKeystoreOpensearchCertificatesProvider.class).asEagerSingleton();
+        opensearchCertificatesProviders.addBinding().to(NoOpensearchCertificatesProvider.class).asEagerSingleton();
+
+
+        //opensearch configuration beans. The order of the beans is important here!
         Multibinder<OpensearchConfigurationBean> opensearchConfigurationBeanMultibinder = Multibinder.newSetBinder(binder(), OpensearchConfigurationBean.class);
+        opensearchConfigurationBeanMultibinder.addBinding().to(OpensearchDefaultConfigFilesBean.class).asEagerSingleton();
+        opensearchConfigurationBeanMultibinder.addBinding().to(OpensearchCommonConfigurationBean.class).asEagerSingleton();
+        opensearchConfigurationBeanMultibinder.addBinding().to(OpensearchCommonConfigurationBean.class).asEagerSingleton();
```
[the remark this answers is not in the document]
Comment (todvora) [reply data-node/src/main/java/org/graylog/datanode/bindings/OpensearchProcessBindings.java]: Thanks, fixed!

File: data-node/src/main/java/org/graylog/datanode/opensearch/configuration/beans/files/ConfigFile.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog.datanode.opensearch.configuration.beans.files;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.file.Path;
+
+public interface ConfigFile {
```
[the remark this answers is not in the document]
Comment (moesterheld) [reply data-node/src/main/java/org/graylog/datanode/opensearch/configuration/beans/files/ConfigFile.java:23]: Actually, I meant moving the `ConfigFile` interface and implementing classes out of the opensearch package. But we can also do this when reusing them for trino.

[the remark this answers is not in the document]
Comment (todvora) [reply data-node/src/main/java/org/graylog/datanode/opensearch/configuration/beans/files/ConfigFile.java]: Makes sense, just pushed the change. It now lives under `org.graylog.datanode.process.configuration`, which feels quite correct.

**Review by moesterheld — APPROVED** [review]
lgtm now. Thank you again for the massive change.

**Comment by todvora** [comment]
Thank you for reviewing, testing and merging! Massive task as well!

**Pull request merged**